### ISSUE TITLE
Choose different hook for poetry-tracking-mode

### ISSUE THREAD

**Issue by rdugan** [galaunay/poetry.el#14]
The use of `post-command-hook` for `poetry-tracking-mode` causes the feature to be unusable for me, as every keystroke triggers the `poetry-track-virtualenv` function which slows editing to a crawl.  

The `python-mode-hook` and `projectile-after-switch-project-hook` (assuming projectile) may be more appropriate.

The following `use-package` statement is a hack to get around the issue (while also auto enabling the feature on the hooks mentioned)

```
(use-package poetry
  :ensure t
  :config
  (add-hook 'poetry-tracking-mode-hook (lambda () (remove-hook 'post-command-hook 'poetry-track-virtualenv)))
  (add-hook 'python-mode-hook 'poetry-track-virtualenv)
  (add-hook 'projectile-after-switch-project-hook 'poetry-track-virtualenv))
```

**Comment by galaunay** [comment]
I agree it can be pretty slow.
Unfortunately, I am not aware of any other generic solution.

The best we can do is leave the choice to the user (between using 'post-command-hook and 'projectile-after-switch-project-hook)

**Comment by gergelypolonkai** [comment]
How about putting it to `python-mode-hook` by default, and either adding it to `projectile-after-switch-project-hook` if Projectile is installed, or documenting that the user can (and maybe should?) do it.

**Comment by rdugan** [comment]
For anyone looking for at least a start at a faster but somewhat robust solution, here's the trickery I ended up with after some experimentation.  This is to get around the fact that python mode doesn't cover all python project files (e.g. toml), as well as the fact that projectile-after-switch-project-hook only runs on manual switches:

```
(use-package poetry
  :ensure t
  :config
  (add-hook 'find-file-hook
            (lambda () (poetry-track-virtualenv)))
  (advice-add 'kill-buffer
              :around
              (lambda (func &rest args)
                (let* ((next-buffer-name (buffer-file-name (other-buffer)))
                       (both-file-buffers (and next-buffer-name buffer-file-name)))
                  (apply func args)
                  (if both-file-buffers (poetry-track-virtualenv)))))
  (advice-add 'switch-to-buffer
              :after
              (lambda (&rest args)
                (if buffer-file-name (poetry-track-virtualenv))))
  (advice-add 'windmove-do-window-select
              :after
              (lambda (&rest args)
                (if buffer-file-name (poetry-track-virtualenv))))
  )
```

First hook is for file opens, second is for buffer switch after kill, third is for a manual buffer switch, and fourth is for moving between two visible buffers (windows.) This is probably fragile, and leaves some use cases uncovered, but seems to address most of my everyday scenarios.

**Comment by revmischa** [comment]
I simply added `poetry` to spacemacs, which uses projectile by default. Unfortunately yeah, doing literally anything causes it to pause and emit "Compilation finished" and "Poetry finished". This makes poetry-tracking-mode unusable with spacemacs :/

**Comment by galaunay** [comment]
You are right that it can be quite unusable in its current state.
I made a PR to add the possibility to use projectile or @rdugan solution.

I just tested it quickly, it would be nice to know if it works for you before going forward.

**Comment by galaunay** [comment]
I did more complete tests and merged #18 .

You can now use projectile with  `(setq poetry-tracking-strategy 'projectile)` and @rdugan method (thanks for the code) with `(setq poetry-tracking-strategy 'switch-buffer)`.

Please re-open if I missed something.

**Comment by rdugan** [comment]
Sorry for the late response - I've definitely seen some quirkiness in my setup - specifically some kind of infinite recursion or something of the sort.  Can't say for certain that it's due to the code I pasted above, but that is the most likely suspect.  I would suggest documenting that as experimental if merging it into your codebase.  Cheers!
